operator/internal/controller/redpanda/role_controller_test.go

```diff
@@ -542,3 +551,153 @@ func TestRoleLifecycleTransitions(t *testing.T) {
+	defer cancel()
+
+	timeoutOption := kgo.RetryTimeout(1 * time.Millisecond)
+	environment := InitializeResourceReconcilerTest(t, ctx, &RoleReconciler{
+		extraOptions: []kgo.Opt{timeoutOption},
+	})
+
+	// Create a role with initial members
+	role := &redpandav1alpha2.RedpandaRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: metav1.NamespaceDefault,
+			Name:      "membership-role-" + strconv.Itoa(int(time.Now().UnixNano())),
+		},
+		Spec: redpandav1alpha2.RoleSpec{
+			ClusterSource: environment.ClusterSourceValid,
+			Principals:    []string{"User:alice", "User:bob"},
+		},
+	}
+
+	key := client.ObjectKeyFromObject(role)
+	req := ctrl.Request{NamespacedName: key}
+
+	// Initial creation
+	t.Run("initial_creation", func(t *testing.T) {
+		require.NoError(t, environment.Factory.Create(ctx, role))
+		_, err := environment.Reconciler.Reconcile(ctx, req)
+		require.NoError(t, err)
+
+		require.NoError(t, environment.Factory.Get(ctx, key, role))
+		require.True(t, role.Status.ManagedRole)
+		require.True(t, role.Status.ManagedPrincipals)
+		require.Equal(t, []string{"User:alice", "User:bob"}, role.Spec.Principals)
+
+		// Verify role exists with correct members
+		rolesClient, err := environment.Factory.Roles(ctx, role)
+		require.NoError(t, err)
+		defer rolesClient.Close()
+
+		hasRole, err := rolesClient.Has(ctx, role)
+		require.NoError(t, err)
+		require.True(t, hasRole)
+	})
+
+	// Test 1: Add a new member
+	t.Run("add_member", func(t *testing.T) {
+		require.NoError(t, environment.Factory.Get(ctx, key, role))
+		role.Spec.Principals = []string{"User:alice", "User:bob", "User:charlie"}
+
+		require.NoError(t, environment.Factory.Update(ctx, role))
+		_, err := environment.Reconciler.Reconcile(ctx, req)
+		require.NoError(t, err)
+
+		require.NoError(t, environment.Factory.Get(ctx, key, role))
+		require.True(t, role.Status.ManagedRole)
+		require.True(t, role.Status.ManagedPrincipals)
+		require.Equal(t, []string{"User:alice", "User:bob", "User:charlie"}, role.Spec.Principals)
+
+		// Verify the role still exists and reconciliation was triggered
+		rolesClient, err := environment.Factory.Roles(ctx, role)
+		require.NoError(t, err)
+		defer rolesClient.Close()
+
+		hasRole, err := rolesClient.Has(ctx, role)
+		require.NoError(t, err)
+		require.True(t, hasRole, "Role should still exist after membership update")
+	})
+
+	// Test 2: Remove a member
+	t.Run("remove_member", func(t *testing.T) {
+		require.NoError(t, environment.Factory.Get(ctx, key, role))
+		role.Spec.Principals = []string{"User:alice", "User:charlie"}
+
+		require.NoError(t, environment.Factory.Update(ctx, role))
+		_, err := environment.Reconciler.Reconcile(ctx, req)
+		require.NoError(t, err)
+
+		require.NoError(t, environment.Factory.Get(ctx, key, role))
+		require.True(t, role.Status.ManagedRole)
+		require.True(t, role.Status.ManagedPrincipals)
+		require.Equal(t, []string{"User:alice", "User:charlie"}, role.Spec.Principals)
```

Comment:
I feel like comparison with role.Spec.Principals does not test reconciler, but rather client. Shouldn't you get principals from Redpanda and compare?